src/sas/sascalc/corfunc/corfunc_calculator.py

```diff
@@ -1,290 +1,670 @@
+        if self._porod is None:
+            return None
+
+        return self._porod.data
+
+    @porod.setter
+    def porod(self, value: Optional[PorodData]):
+        self._porod.data = value
 
-        # Ensure the errors are set correctly
-        if new_data.dy is None or len(new_data.x) != len(new_data.dy) or \
-            (min(new_data.dy) == 0 and max(new_data.dy) == 0):
-            new_data.dy = np.ones(len(new_data.x))
+    @property
+    def transformed(self):
+        return self._transformed_data
+
+    @property
+    def lamellar_parameters(self) -> LamellarParameters:
+        return self._lamellar_parameters
+
+    @property
+    def supplementary_parameters(self) -> Optional[SupplementaryParameters]:
+        return self._supplementary_parameters
+
+    @property
+    def extrapolation_function(self) -> Optional[Callable]:
+        return self._extrapolation_function
+
+    @property
+    def min_extrapolated(self) -> Optional[float]:
+        if self._extrapolation_data is None:
+            return None
+        else:
+            return np.min(self._extrapolation_data.y)
 
-        self._data = new_data
+    @property
+    def fit_background(self):
+        return self._background.allow_fit
+
+    @fit_background.setter
+    def fit_background(self, value: bool):
+        self._background.allow_fit = value
+
+    @property
+    def fit_guinier(self):
+        return self._guinier.allow_fit
+
+    @fit_guinier.setter
+    def fit_guinier(self, value: bool):
+        self._guinier.allow_fit = value
+
+    @property
+    def fit_porod(self):
+        return self._porod.allow_fit
+
+    @fit_porod.setter
+    def fit_porod(self, value: bool):
+        self._porod.allow_fit = value
+
+    @property
+    def extrapolated(self):
+        return self._extrapolation_data
 
-    def compute_background(self, upperq=None):
+    #
+    # Calculation Steps
+    #
+
+
+    def run(self):
+        """ Execute the calculation"""
+        self._calculate_background()
+        self._calculate_background_subtracted()
+        self._calculate_porod_parameters()
+        self._calculate_guinier_parameters()
+        self._calculate_extrapolation_function()
+        self._calculate_extrapolation_data()
+        self._calculate_transforms()
+        self._calculate_parameters()
+
+
+
+    # Calculation components
+
+    def _calculate_background(self):
         """
-        Compute the background level from the Porod region of the data
+        Compute the background level from the Porod region of the data,
+        only do this if background fitting is allowed
         """
-        if self._data is None: return 0
-        elif upperq is None and self.upperq is not None: upperq = self.upperq
-        elif upperq is None and self.upperq is None: return 0
-        q = self._data.x
-        mask = np.logical_and(q > upperq[0], q < upperq[1])
-        _, _, bg = self._fit_porod(q[mask], self._data.y[mask])
+        if self._data is None:
```

Comment:
It can be abstracted out, but I generally try to avoid that kind of abstraction in favour of readability and standardisation... that is, if I can be arsed to type it all out in long form.